Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -79,6 +79,14 @@ instance : AddCancelMonoid DyckWord where
   add_left_cancel p q r h := by rw [DyckWord.ext_iff] at *; exact append_cancel_left h
   add_right_cancel p q r h := by rw [DyckWord.ext_iff] at *; exact append_cancel_right h
 
+/-- We define a partial order on Dyck words by writing `p ≤ q` if `p` is a suffix of `q`. -/
+instance : PartialOrder DyckWord where
+  le p q := p.toList <:+ q.toList
+  le_refl p := suffix_rfl
+  le_trans p q r := IsSuffix.trans
+  le_antisymm p q pq qp :=
+    DyckWord.ext_iff.mpr <| eq_of_suffix_of_length_eq pq <| pq.length_le.antisymm qp.length_le
```

Comment:
Can you provide lemmas like ` p.toList <:+ q.toList ↔ p < q`, `0 < p ↔ p ≠ 0`, ...? I suspect it might have to be a separate PR

Reply:
I've probably added – and changed – enough in this PR; we can merge this in its current state and leave the extra ordering lemmas to the next and last PR in my plans, #9781 itself where I will add the tree bijection/relation to Catalan numbers.

Reply:
Sure

maintainer merge